docx/oxml/numbering.py

```diff
@@ -152,30 +152,93 @@ def get_num_for_p(self, p, styles_cache):
         """
         Returns list item for the given paragraph.
         """
-        numPr = p.pPr.get_numPr(styles_cache)
-        ilvl, numId = numPr.ilvl, numPr.numId.val
-        ilvl = ilvl.val if ilvl is not None else 0
+
+        def get_ilvl_and_numId(paragraph):
+            """
+            Return ``ilvl`` and ``numId`` for the given ``paragraph``.
+            """
+            para_numPr = paragraph.pPr.get_numPr(styles_cache)
+            para_ilvl, para_numId = para_numPr.ilvl, para_numPr.numId.val
+            para_ilvl = para_ilvl.val if para_ilvl is not None else 0
+            return para_ilvl, para_numId
+
+        def get_preceding_paragraphs_numIds(p, p_ilvl, p_numId):
+            """
+            Return preceding siblings ``numId`` that are in the same numbered list as the paragraph ``p``.
+            Paragraphs are in the same list if they are on the same level (``p_ilvl``), and either
+            have the same ``p_numId`` or have the same ``pStyle.val``.
+            Skips unnumbered paragraphs within the numbering list.
+            Stops on the paragraph that is on the lower level.
+            """
+            pStyle = p.pPr.pStyle
+            for para in p.itersiblings(preceding=True):
+                try:
+                    para_ilvl, para_numId = get_ilvl_and_numId(para)
+                    # skip unnumbered paragraphs within numbering list
+                    if para_numId == 0:
+                        continue
+                    paStyle = para.pPr.pStyle
+                    if para_ilvl < p_ilvl and (para_numId == p_numId or pStyle.val == paStyle.val or
+                                         (paStyle is not None and paStyle.val in linked_styles)):
+                        break
+                    if para_ilvl == p_ilvl and (para_numId == p_numId or pStyle.val == paStyle.val or paStyle.val in linked_styles):
+                        yield para_numId
+                except AttributeError:
+                    continue
+                except StopIteration:
+                    break
+
+        def count_same_numIds(preceding_paragraphs_numIds, numId, num):
```

Comment:
i was wondering if we can reuse this for loop https://github.com/openlawlibrary/python-docx/pull/92/files#diff-7d408ea4319091551a0ad299ce348a61450094d1f5e7386e465903c461202b11R165 since we're grabbing `numId` of each previous para on this line https://github.com/openlawlibrary/python-docx/pull/92/files#diff-7d408ea4319091551a0ad299ce348a61450094d1f5e7386e465903c461202b11R176 and integrate this numbering count in it, that might reduce complexity since we're reusing existing for instead of having it been executed twice. maybe this is something we can create an issue for, for some future refactoring?